### PR TITLE
Extend filtering

### PR DESCRIPTION
**Pull request by brkgbr**
Extended article filtering and implemented state saving for QToolButtons
Article filtering

### CODE REVIEW

File: src/librssguard/core/messagesproxymodel.h [hunk 1]
```diff
@@ -27,19 +42,21 @@ class MessagesProxyModel : public QSortFilterProxyModel {
...
-    bool showUnreadOnly() const;
-    void setShowUnreadOnly(bool show_unread_only);
+    void setFilter(MessageFilter filter);
 
   private:
     QModelIndex getNextImportantItemIndex(int default_row, int max_row) const;
     QModelIndex getNextUnreadItemIndex(int default_row, int max_row) const;
 
     bool lessThan(const QModelIndex& left, const QModelIndex& right) const;
+    bool acceptMessage(Message currentMessage) const;
```
Comment (martinrotter) [inline src/librssguard/core/messagesproxymodel.h:52]: Method's name is not self-explanatory enough, perhaps add word "filter" somewhere in method's name.

File: src/librssguard/core/messagesproxymodel.cpp [hunk 2]
```diff
@@ -100,6 +100,44 @@ bool MessagesProxyModel::lessThan(const QModelIndex& left, const QModelIndex& ri
   return false;
 }
 
+bool MessagesProxyModel::acceptMessage(Message currentMessage) const {
+  const QDate currentDate = QDate::currentDate();
```
Comment (martinrotter) [inline src/librssguard/core/messagesproxymodel.cpp:104]: These two lines are only used in some switch cases, rearrange the method to only make those calls when needed.

Usually these methods which are used for repeated calling (especially in "proxy" and "source" models) must be optimized as much as possible. Imagine displaying 10 000 articles - which is not rare in RSS Guard - and calling `const QDateTime currentDateTime = QDateTime::currentDateTime();` uselessly 10 000 times. Also only call `currentDateTime` and then call `.date()` on it, it's bit faster.

File: src/librssguard/core/messagesproxymodel.cpp [hunk 3]
```diff
@@ -11,7 +11,7 @@
 #include <QTimer>
 
 MessagesProxyModel::MessagesProxyModel(MessagesModel* source_model, QObject* parent)
-  : QSortFilterProxyModel(parent), m_sourceModel(source_model), m_showUnreadOnly(false) {
+  : QSortFilterProxyModel(parent), m_sourceModel(source_model), m_filter(MessageFilter::NoFiltering) {
```
Comment (martinrotter) [inline src/librssguard/core/messagesproxymodel.cpp:14]: I suggest renaming the enum to `MessageListFilter` as there already is `MessageFilter` class for a different feature.

File: src/librssguard/gui/toolbars/messagestoolbar.cpp [hunk 4]
```diff
@@ -125,26 +140,128 @@ void MessagesToolBar::initializeSearchBox() {
   });
 }
 
+void MessagesToolBar::addActionToMenu(QMenu* menu, const QIcon& icon, const QString& title, const QVariant& value, const QString& name) {
```
Comment (martinrotter) [inline src/librssguard/gui/toolbars/messagestoolbar.cpp:143]: Good catch.

File: src/librssguard/gui/toolbars/messagestoolbar.cpp [hunk 5]
```diff
@@ -125,26 +140,128 @@ void MessagesToolBar::initializeSearchBox() {
...
+  m_actionMessageFilter->setDefaultWidget(m_btnMessageFilter);
+  m_actionMessageFilter->setIcon(m_btnMessageFilter->icon());
+  m_actionMessageFilter->setProperty("type", FILTER_ACTION_NAME);
+  m_actionMessageFilter->setProperty("name", tr("Article filter"));
 
   connect(m_menuMessageHighlighter, &QMenu::triggered, this, &MessagesToolBar::handleMessageHighlighterChange);
+  connect(m_menuMessageFilter, &QMenu::triggered, this, &MessagesToolBar::handleMessageFilterChange);
+}
+
+void MessagesToolBar::saveToolButtonSelection(const QString& buttonName, const QAction* action) const {
```
Comment (martinrotter) [inline src/librssguard/gui/toolbars/messagestoolbar.cpp:241]: I always use "snake case" notation for local variables, so `button_name` except of `buttonName`, but for class MEMBER variables, I use "pascal notation" with `m_` prefix.

File: src/librssguard/gui/toolbars/messagestoolbar.cpp [hunk 6]
```diff
@@ -125,26 +140,128 @@ void MessagesToolBar::initializeSearchBox() {
...
+  QStringList action_names = savedActions();
+  for (QString& action_name : action_names) {
+    if (action_name.startsWith(buttonName)) {
+      action_name = buttonName + (action->objectName().isEmpty() ? "" : "[" + action->objectName().toStdString() + "]").c_str();
+    }
+  }
+  qApp->settings()->setValue(GROUP(GUI), GUI::MessagesToolbarDefaultButtons, action_names.join(QSL(",")));
+}
+
+void MessagesToolBar::activateAction(const QString& action_name, QWidgetAction* widgetAction) {
```
Comment (martinrotter) [inline src/librssguard/gui/toolbars/messagestoolbar.cpp:251]: Paramater name not in "snake notation".

File: src/librssguard/gui/toolbars/messagestoolbar.cpp [hunk 7]
```diff
@@ -125,26 +140,128 @@ void MessagesToolBar::initializeSearchBox() {
...
+  const int start = action_name.indexOf('[');
+  const int end = action_name.indexOf(']');
```
Comment (martinrotter) [inline src/librssguard/gui/toolbars/messagestoolbar.cpp:253]: I format my code with "uncrustify" with config found in [uncrustify.cfg](https://github.com/martinrotter/rssguard/blob/master/resources/scripts/uncrustify/uncrustify.cfg)

**Review by martinrotter — CHANGES_REQUESTED** [review]

File: src/librssguard/core/messagesproxymodel.h (hunk 1 above)
Comment (brkgbr) [reply src/librssguard/core/messagesproxymodel.h]: Changed it and thanks

File: src/librssguard/gui/toolbars/messagestoolbar.cpp (hunk 5 above)
Comment (brkgbr) [reply src/librssguard/gui/toolbars/messagestoolbar.cpp]: Yeah I tried to follow the conventions, I think I missed this one, will follow it from now on

File: src/librssguard/gui/toolbars/messagestoolbar.cpp (hunk 6 above)
Comment (brkgbr) [reply src/librssguard/gui/toolbars/messagestoolbar.cpp]: Done

**Comment by akinokonomi** [comment]
Show articles from today -> Show today's articles
Show articles from yesterday -> Show yesterday's articles
Show articles from last 24 hours -> Show articles in last 24 hours
Show articles from last 48 hours -> Show articles in last 48 hours

May also change:

Show articles from this week -> Show this week articles
Show articles from last week -> Show last week articles

**Review by martinrotter — APPROVED** [review]
I am impressed. This is good PR and although I will do some cosmetic changes, it is good.

Thanks, merging.

File: src/librssguard/gui/toolbars/messagestoolbar.cpp (hunk 7 above)
Comment (martinrotter) [reply src/librssguard/gui/toolbars/messagestoolbar.cpp:256]: I will check if clang-format is good for me and will switch as it is indeed much more used than uncrustify these times.

Comment (martinrotter) [reply src/librssguard/gui/toolbars/messagestoolbar.cpp:256]: OK, I switched to clang-format.

https://github.com/martinrotter/rssguard/blob/master/.clang-format

I will gradually update the config file to make it work for me and will re-format all source files with it eventually.

**Comment by amreo** [comment]
Mhm where is the QtoolButtons in rssguard 4.2.2?

**Comment by martinrotter** [comment]
> Mhm where is the QtoolButtons in rssguard 4.2.2?

https://github.com/martinrotter/rssguard/issues/715

**Pull request merged**